File: 프로그래머스/lv3/118668. 코딩 테스트 공부/코딩 테스트 공부.py
#알고력과 코딩력
#공부는 알고력과 코딩력을 각각 1씩 높임
#같은 문제 여러번 푸는 것 가능, 각 문제를 풀면 올라가는 알고력과 코딩력이 정해져있음
#모든 문제를 풀 수 있는 알고력과 코딩력을 얻는 최단 시간을 구하려함

import heapq

def solution(alp, cop, problems):
    answer = 0
    
    #힙 초기화
    heap = []
    heapq.heappush(heap, [0,alp,cop])
    INF = int(1e9)
    dist = [[INF]*151 for i in range(151)]
    true = [[False]*151 for i in range(151)]
    dist[alp][cop] = 0

    while heap:
        k, x, y = heapq.heappop(heap)
        
        if true[x][y] == True:
            continue
        
        true[x][y] = True
        check = True
        
        #문제를 풀 수 있다면 풀기
        for a, b, c, d, e in problems:
            nx = min(x+c,150)
            ny = min(y+d,150)
            
            if a <= x and b <= y:
                if true[nx][ny] == False:
                    if dist[nx][ny] > k+e:
                        heapq.heappush(heap, [k+e, nx, ny])
            else:
                check = False
        
        #모든 문제를 풀 수 있다면
        if check == True:
            answer = k
            break
        
        nx = min(x+1,150)
        if true[nx][y] == False:
            if dist[nx][y] > k+1:
                heapq.heappush(heap, [k+1, nx, y])
        
        ny = min(y+1,150)
        if true[x][ny] == False:
            if dist[x][ny] > k+1:
                heapq.heappush(heap, [k+1, x, ny])
    
    return answer

File: 프로그래머스/lv3/118668. 코딩 테스트 공부/코딩 테스트 공부.java
import java.util.*;

class Node {
    
    int d; //거리
    int al; //알고력
    int co; //코딩력
    
    //생성자
    public Node(int d, int al, int co) {
        this.d = d;
        this.al = al;
        this.co = co;
    }
}

class Solution {

    public static int k;
    
    public int solution(int alp, int cop, int[][] problems) {
        int answer = 0;
        boolean[][] check = new boolean[151][151];
        int[][] dist = new int[151][151];
        
        int Max = 100000000;
        for (int i=0; i<151; i++) {
            for (int j=0; j<151; j++) {
                dist[i][j] = Max;
            }
        }
        dist[alp][cop] = 0;
        
        PriorityQueue<Node> heap = new PriorityQueue<>((o1, o2) -> {
            return o1.d-o2.d;
        });
        
        
        heap.add(new Node(0,alp,cop));
        
        int d;
        int al;
        int co;
        int nal;
        int nco;
        Node now;
        boolean c;
        //다익스트라 시작
        while (heap.size() > 0) {
            now = heap.poll();
            d = now.d;
            al = now.al;
            co = now.co;
            
            
            if (check[al][co] == true) {
                continue;
            }
            
            check[al][co] = true;
            
            c = true;
            //모든 경우 다 할 수 있는지 파악
            for (int[] problem : problems) {
                //풀수 있다면
                if ((problem[0] <= al) && (problem[1] <= co)) {
                    nal = Math.min(al+problem[2], 150);
                    nco = Math.min(co+problem[3], 150);
                    
                    if (check[nal][nco] == false) {
                        
                        if (dist[nal][nco] > d + problem[4]) {
                            dist[nal][nco] = d + problem[4];
                            heap.add(new Node(dist[nal][nco], nal, nco));
                        }
                    }

                } else {
                    c = false;
                }
            }
            
            if (c == true) {
                answer = d;
                return answer;
            }
            
            nal = Math.min(al+1, 150);
            nco = Math.min(co+1, 150);

            if (check[nal][co] == false) {

                if (dist[nal][co] > d + 1) {
                    dist[nal][co] = d + 1;
                    heap.add(new Node(dist[nal][co], nal, co));
                }
            }
            
            if (check[al][nco] == false) {

                if (dist[al][nco] > d + 1) {
                    dist[al][nco] = d + 1;
                    heap.add(new Node(dist[al][nco], al, nco));
                }
            }
        }
        
        
        
        return answer;
    }
}